app/src/main/java/com/babylon/wallet/android/presentation/settings/securitycenter/securityfactors/biometricspin/BiometricsPinViewModel.kt

```diff
@@ -0,0 +1,176 @@
+package com.babylon.wallet.android.presentation.settings.securitycenter.securityfactors.biometricspin
+
+import androidx.lifecycle.viewModelScope
+import com.babylon.wallet.android.di.coroutines.DefaultDispatcher
+import com.babylon.wallet.android.presentation.common.OneOffEvent
+import com.babylon.wallet.android.presentation.common.OneOffEventHandler
+import com.babylon.wallet.android.presentation.common.OneOffEventHandlerImpl
+import com.babylon.wallet.android.presentation.common.StateViewModel
+import com.babylon.wallet.android.presentation.common.UiState
+import com.babylon.wallet.android.presentation.ui.model.factors.FactorSourceCard
+import com.babylon.wallet.android.presentation.ui.model.factors.FactorSourceStatusMessage
+import com.babylon.wallet.android.utils.relativeTimeFormatted
+import com.radixdlt.sargon.Account
+import com.radixdlt.sargon.DeviceFactorSource
+import com.radixdlt.sargon.EntitiesLinkedToFactorSource
+import com.radixdlt.sargon.FactorSource
+import com.radixdlt.sargon.FactorSourceFlag
+import com.radixdlt.sargon.FactorSourceId
+import com.radixdlt.sargon.FactorSourceIntegrity
+import com.radixdlt.sargon.Persona
+import com.radixdlt.sargon.ProfileToCheck
+import com.radixdlt.sargon.extensions.asGeneral
+import com.radixdlt.sargon.extensions.id
+import com.radixdlt.sargon.extensions.kind
+import com.radixdlt.sargon.os.SargonOsManager
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.collections.immutable.PersistentList
+import kotlinx.collections.immutable.persistentListOf
+import kotlinx.collections.immutable.toPersistentList
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.filterIsInstance
+import kotlinx.coroutines.flow.firstOrNull
+import kotlinx.coroutines.flow.flatMapConcat
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import rdx.works.core.preferences.PreferencesManager
+import rdx.works.profile.domain.GetProfileUseCase
+import javax.inject.Inject
+
+@OptIn(ExperimentalCoroutinesApi::class)
+@HiltViewModel
+class BiometricsPinViewModel @Inject constructor(
+    private val sargonOsManager: SargonOsManager,
+    private val preferencesManager: PreferencesManager,
+    getProfileUseCase: GetProfileUseCase,
+    @DefaultDispatcher private val defaultDispatcher: CoroutineDispatcher
+) : StateViewModel<BiometricsPinViewModel.State>(),
+    OneOffEventHandler<BiometricsPinViewModel.Event> by OneOffEventHandlerImpl() {
+
+    override fun initialState(): State = State()
+
+    init {
+        getProfileUseCase.flow
+            .flatMapConcat { profile ->
```

Comment:
This caught my attention and it took me a minute to understand what happens here. I'm not against keeping it as it is, but maybe it's a sign that the readability can be a little bit improved. What happens now is that each list element emits and for each element there are some operations performed below. We would benefit a lot from this approach if there were any async operations performed, because we'd have the UI updated gradually instead of waiting for all the operations, but it's not the case because the only async operation I can see is performed is reading from prefs which is very fast anyway. 

No need to change anything, it's just a thought.